src/app.ts

```diff
@@ -9,20 +9,46 @@ import type { FastifyRequest, FastifyReply } from 'fastify'
 
 const w = usb()
 
+const validSubprotocols: string[] = ['XAPWS1']
+
 const app = fastify({ logger: true })
-app.register(fastifyws)
-
-app.get('/ws', { websocket: true }, (connection: SocketStream, req: FastifyRequest) => {
-  connection.socket.on('message', (message: MessageEvent) => {
-    w.postMessage(message.toString())
-    w.once('message', (result) => {
-      req.log.info(result)
-      connection.socket.send(result)
-    })
-    w.on('error', (err) => {
-      req.log.error(err)
-      connection.socket.send(err.toString())
-      process.exit(1)
+app.register(fastifyws, {
+  options: {
+    /**
+     * Specify how to format sec-websocket-protocol header on upgrade
+     * @param protocols set of subprotocols from header
+     * @returns subprotocol accepted or false
+     */
+    handleProtocols: (protocols: Set<string>) => {
+      if (protocols.has('XAPWS1')) {
+        return 'XAPWS1'
+      }
+      return false
+    }
+  }
+})
+
+app.register(async (fastify: FastifyInstance) => {
+  fastify.addHook('preValidation', async (req: FastifyRequest, reply: FastifyReply) => {
```

Comment:
With the comment on `handleProtocols` we can get rid of `prevalidation` hook.
also you don't need to await the `reply.send()` just return it.

Reply:
Sounds good